src/share/man/man3/sysexits.3
.\"
.\" Copyright (c) 1996 Joerg Wunsch
.\"
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE DEVELOPERS ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE DEVELOPERS BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" $OpenBSD: sysexits.3,v 1.6 2003/05/20 09:13:38 jmc Exp $
.\" $FreeBSD: src/share/man/man3/sysexits.3,v 1.4.2.1 1999/08/29 16:45:41 peter Exp $
.\"
.Dd March 31, 1996
.Dt SYSEXITS 3
.Os
.Sh NAME
.Nm sysexits
.Nd preferable exit codes for programs
.Sh SYNOPSIS
.Fd #include <sysexits.h>
.Sh DESCRIPTION
According to
.Xr style 9 ,
it is not good practice to call
.Xr exit 3
with arbitrary values to indicate a failure condition when ending a program.
Instead, the pre-defined exit codes from
.Nm
should be used, so the caller of the process can get a rough
estimation about the failure class without looking up the source code.
.Pp
The successful exit is always indicated by a status of 0, or
.Dv EX_OK .
Error numbers begin at
.Dv EX__BASE
to reduce the possibility of clashing with other exit statuses that
random programs may already return.
The meaning of the code is approximately as follows:
.Bl -tag -width "EX_UNAVAILABLEXX(XX)"
.It Dv EX_USAGE Pq 64
The command was used incorrectly, e.g., with the wrong number of
arguments, a bad flag, bad syntax in a parameter, or whatever.
.It Dv EX_DATAERR Pq 65
The input data was incorrect in some way.
This should only be used for user's data and not system files.
.It Dv EX_NOINPUT Pq 66
An input file (not a system file) did not exist or was not readable.
This could also include errors like
.Dq \&No message
to a mailer (if it cared to catch it).
.It Dv EX_NOUSER Pq 67
The user specified did not exist.
This might be used for mail addresses or remote logins.
.It Dv EX_NOHOST Pq 68
The host specified did not exist.
This is used in mail addresses or network requests.
.It Dv EX_UNAVAILABLE Pq 69
A service is unavailable.
This can occur if a support program or file does not exist.
This can also be used as a catch-all message when something you wanted
to do doesn't work, but you don't know why.
.It Dv EX_SOFTWARE Pq 70
An internal software error has been detected.
This should be limited to non-operating system related errors if possible.
.It Dv EX_OSERR Pq 71
An operating system error has been detected.
This is intended to be used for such things as
.Dq cannot fork ,
or
.Dq cannot create pipe .
It includes things like
.Xr getuid 2
returning a user that does not exist in the passwd file.
.It Dv EX_OSFILE Pq 72
Some system file (e.g.,
.Pa /etc/passwd ,
.Pa /var/run/utmp )
does not exist, cannot be opened, or has some sort of error
(e.g., syntax error).
.It Dv EX_CANTCREAT Pq 73
A (user specified) output file cannot be created.
.It Dv EX_IOERR Pq 74
An error occurred while doing I/O on some file.
.It Dv EX_TEMPFAIL Pq 75
Temporary failure, indicating something that is not really an error.
In sendmail, this means that a mailer, for example, could not create a
connection, and the request should be reattempted later.
.It Dv EX_PROTOCOL Pq 76
The remote system returned something that was
.Dq not possible
during a protocol exchange.
.It Dv EX_NOPERM Pq 77
You did not have sufficient permission to perform the operation.
This is not intended for filesystem problems, which should use
.Dv EX_NOINPUT
or
.Dv EX_CANTCREAT ,
but rather for higher level permissions.
.It Dv EX_CONFIG Pq 78
Something was found in an unconfigured or misconfigured state.
.El
.Pp
The numerical values corresponding to the symbolical ones are given in
parentheses for easy reference.
.Sh SEE ALSO
.Xr exit 3 ,
.Xr style 9
.Sh HISTORY
The
.Nm
file appeared sometime after
.Bx 4.3 .
.Sh AUTHORS
This man page was written by
.ie t J\(:org Wunsch
.el Joerg Wunsch
after the comments in
.Aq Pa sysexits.h .
.Sh BUGS
The choice of an appropriate exit value is often ambiguous.


src/gnu/usr.sbin/sendmail/sendmail/sysexits.c
/*
 * Copyright (c) 1998-2001 Proofpoint, Inc. and its suppliers.
 *	All rights reserved.
 * Copyright (c) 1983, 1995-1997 Eric P. Allman.  All rights reserved.
 * Copyright (c) 1988, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * By using this file, you agree to the terms and conditions set
 * forth in the LICENSE file which can be found at the top level of
 * the sendmail distribution.
 *
 */

#include <sendmail.h>

SM_RCSID("@(#)$Id$")

/*
**  DSNTOEXITSTAT -- convert DSN-style error code to EX_ style.
**
**	Parameters:
**		dsncode -- the text of the DSN-style code.
**
**	Returns:
**		The corresponding exit status.
*/

int
dsntoexitstat(dsncode)
	char *dsncode;
{
	int code2, code3;

	/* first the easy cases.... */
	if (*dsncode == '2')
		return EX_OK;
	if (*dsncode == '4')
		return EX_TEMPFAIL;

	/* reject other illegal values */
	if (*dsncode != '5')
		return EX_CONFIG;

	/* now decode the other two field parts */
	if (*++dsncode == '.')
		dsncode++;
	code2 = atoi(dsncode);
	while (*dsncode != '\0' && *dsncode != '.')
		dsncode++;
	if (*dsncode != '\0')
		dsncode++;
	code3 = atoi(dsncode);

	/* and do a nested switch to work them out */
	switch (code2)
	{
	  case 0:	/* Other or Undefined status */
		return EX_UNAVAILABLE;

	  case 1:	/* Address Status */
		switch (code3)
		{
		  case 0:	/* Other Address Status */
			return EX_DATAERR;

		  case 1:	/* Bad destination mailbox address */
		  case 6:	/* Mailbox has moved, No forwarding address */
			return EX_NOUSER;

		  case 2:	/* Bad destination system address */
		  case 8:	/* Bad senders system address */
			return EX_NOHOST;

		  case 3:	/* Bad destination mailbox address syntax */
		  case 7:	/* Bad senders mailbox address syntax */
			return EX_USAGE;

		  case 4:	/* Destination mailbox address ambiguous */
			return EX_UNAVAILABLE;

		  case 5:	/* Destination address valid */
			/* According to RFC1893, this can't happen */
			return EX_CONFIG;
		}
		break;

	  case 2:	/* Mailbox Status */
		switch (code3)
		{
		  case 0:	/* Other or Undefined mailbox status */
		  case 1:	/* Mailbox disabled, not accepting messages */
		  case 2:	/* Mailbox full */
		  case 4:	/* Mailing list expansion problem */
			return EX_UNAVAILABLE;

		  case 3:	/* Message length exceeds administrative lim */
			return EX_DATAERR;
		}
		break;

	  case 3:	/* System Status */
		return EX_OSERR;

	  case 4:	/* Network and Routing Status */
		switch (code3)
		{
		  case 0:	/* Other or undefined network or routing stat */
			return EX_IOERR;

		  case 1:	/* No answer from host */
		  case 3:	/* Routing server failure */
		  case 5:	/* Network congestion */
			return EX_TEMPFAIL;

		  case 2:	/* Bad connection */
			return EX_IOERR;

		  case 4:	/* Unable to route */
			return EX_PROTOCOL;

		  case 6:	/* Routing loop detected */
			return EX_CONFIG;

		  case 7:	/* Delivery time expired */
			return EX_UNAVAILABLE;
		}
		break;

	  case 5:	/* Protocol Status */
		return EX_PROTOCOL;

	  case 6:	/* Message Content or Media Status */
		return EX_UNAVAILABLE;

	  case 7:	/* Security Status */
		return EX_DATAERR;
	}
	return EX_UNAVAILABLE;
}
/*
**  EXITSTAT -- convert EX_ value to error text.
**
**	Parameters:
**		excode -- rstatus which might consists of an EX_* value.
**
**	Returns:
**		The corresponding error text or the original string.
*/

char *
exitstat(excode)
	char *excode;
{
	char *c;
	int i;
	char *exitmsg;

	if (excode == NULL || *excode == '\0')
		return excode;
	i = (int) strtol(excode, &c, 10);
	if (*c != '\0')
		return excode;
	exitmsg = sm_sysexitmsg(i);
	if (exitmsg != NULL)
		return exitmsg;
	return excode;
}
